consensus/src/common.rs

```diff
@@ -0,0 +1,211 @@
+// Copyright 2020 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0
+
+use blake2_rfc::blake2b::blake2b;
+use mina_crypto::hash::BaseHash;
+use mina_rs_base::consensus_state::ConsensusState;
+use mina_rs_base::global_slot::GlobalSlot;
+use mina_rs_base::protocol_state::{Header, ProtocolState};
+use serde_bin_prot::to_writer;
+use std::convert::TryInto;
+
+pub struct ProtocolStateChain(Vec<ProtocolState>);
+
+pub trait Chain<T>
+where
+    T: Header,
+{
+    fn push(&mut self, new: T) -> Result<(), &'static str>;
+    fn top(&self) -> Option<&T>;
+    fn consensus_state(&self) -> Option<&ConsensusState>;
+    fn global_slot(&self) -> Option<&GlobalSlot>;
+    fn epoch_slot(&self) -> Option<u32>;
+    fn length(&self) -> u64;
+    fn last_vrf(&self) -> String;
+    fn state_hash(&self) -> Option<BaseHash>;
+}
+
+impl Chain<ProtocolState> for ProtocolStateChain {
+    fn push(&mut self, new: ProtocolState) -> Result<(), &'static str> {
+        match self.0.len() {
+            0 => (),
+            n => {
+                if new.get_height().0 != self.0[n - 1].get_height().0 + 1 {
+                    return Err("header must have height 1 greater than top");
+                }
+            }
+        }
+
+        self.0.push(new);
+        Ok(())
+    }
+
+    fn top(&self) -> Option<&ProtocolState> {
+        self.0.last()
+    }
+
+    fn consensus_state(&self) -> Option<&ConsensusState> {
+        match self.top() {
```

Comment:
We can replace the whole match with just `self.top().map(|s| &s.body.consensus_state)`

Reply:
updated